_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,140 @@
+---
+layout: singlepage-overview
+title: Binary Compatibility for library authors
+
+discourse: true
+permalink: /tutorials/:title.html
+---
+
+## Introduction
+
+A diverse and comprehensive set of libraries is important to any productive software ecosystem. While it is easy to develop and distribute Scala libraries, good library authorship goes far
+beyond just writing code and publishing them.
+
+In this guide, we will cover the important topic of **Binary Compatibility**:
+
+* How binary incompatibility can cause production failures in your applications
+* How library authors can avoid breaking binary compatibility, and/or convey breakages clearly to library users when they happen
+
+Before we start, first we need to understand how code is compiled and executed on the Java Virtual Machine (JVM).
+
+## The JVM execution model
+
+Code compiled to run on the JVM is compiled to a platform-independent format called **JVM bytecode** and stored in **Class File** format (with `.class` extension) and these class files are stored
+in JAR files. The bytecode is what we refer to as the **Binary** format.
+
+When application or library code is compiled, their bytecode invokes named references of classes/methods from their dependencies instead of including the dependencies' actual bytecode
+(unless inlining is explicitly requested). During runtime, the JVM classloader will search through the provided class files for classes/methods referenced by and invoke them.
```

Comment:
<s>unless inlining is explicitly request</s>. Things to note here:
* You're only talking about Scala's inlining.
* The previous sentence is true always. Bytecode dependencies are always references.
  
For clarity, I would encourage you to devote a new paragraph or section to describe what Scala's inlining is for, and how it breaks binary compatibility.

Reply:
Also, I'd write "When a JVM application is started, the JVM classloads the application's dependencies and the application, in a process typically called (dynamic) linking. This is the process during which the JVm resolves the bytecode references between the application and its dependencies." instead of "During runtime, the JVM classloader will search through the provided class files for classes/methods referenced by and invoke them.".

Reply:
If I'm new to the JVM world in general, then I'd read that as "On startup, JVM will load all your classes and tell you if methods/classes are missing". However that's not true - the exceptions are thrown only when the missing class/method is invoked. I wanted to stress that fact. However I'm all for using the right terminologies

Reply:
But this is not true for classes -- all classes that a given class uses are specified in the constant pool. If you try to load a class whose constant pool references to a class that has failed to resolve, the JVM will fail at that very same moment. You're right that this is not the case for methods, which have a more "lazy" resolution.

Reply:
Is that true? It still seems to delay loading a class until instantiation is required.

I tested with the following code:

```
package foo
import bar.Bar

object Main extends App {
  println("hello")
  val foo = new Foo("hello")
  foo.bar
}

class Foo(value: String) {
  def bar: Bar = new Bar()
}

```
(+ empty class Bar in another file)

If I remove the `foo.bar` call, no exception is thrown despite `Bar` being referenced in the method signature. At least that's how I interpret "a given class uses are specified"

Reply:
You're right @jatcwang, TIL this. I thought lazy loading only applied to methods and not classes.